### DISCUSSION TITLE
`image.mean` causes error: convert: NewUserSymbol 'image' after non-assignment operator at '.mean' @ error/fx.c/TranslateExpression/2664.

### DISCUSSION THREAD

**Discussion by PenghaiZhang** [ImageMagick/ImageMagick#6060] [Help] [answered]
### ImageMagick version

7.1.0-52

### Operating system, version and so on

Docker image based on Alpine Linux

### Description

Hi, we have been using ImageMagick to generate thumbnails for years and it works very well when we use Ubuntu. 
But recently we have seen an error about `image.mean` when we run our application in an Alpine Linux based docker image.

The error is: 
`convert: NewUserSymbol 'image' after non-assignment operator at '.mean' @ error/fx.c/TranslateExpression/2664.`

The command we run is
`convert xxx.jpg -threshold 99% -format "%[fx:100*image.mean]" info:`

After lots of investigating I still cannot find out why this happens.

Can I have any help please ? 

Thank you!

**Comment by snibgo** [comment, marked as answer]
Instead of `image.mean`, use just `mean`. I suggest also using `magick` instead of `convert`.
```
magick xxx.jpg -threshold 99% -format "%[fx:100*mean]" info:
```
"`%[fx:...]`" and "`-fx`" have been entirely re-written, retaining most but not all backwards compatability. The new fx is documented at [New FX](http://im.snibgo.com/fxnew.htm).

**Reply by PenghaiZhang** [reply]
Thanks mate! I will give a try!

**Reply by PenghaiZhang** [reply]
This is what I found from the New FX page.

The old "-fx" accepts image provided it is qualified by an image attribute, such as image.mean. The documentation doesn't specify the semantics, but experiments show it is equivalent to s.mean. The new "-fx2" does not accept image as a keyword.

Thanks @snibgo

**Comment by fmw42** [comment]
See https://imagemagick.org/script/fx.php were it shows:

![Screen Shot 2023-02-14 at 2 19 14 PM](https://user-images.githubusercontent.com/13458218/218876200-aff497c1-fbd1-43e2-bdde-2a82b82c8ef6.png)

By image, I think it means u, v, or u[X], where u or u[0] is the first image in the sequence. So

```
magick lena.png -format "%[fx:u.mean]" info:
0.502798
```

or just

```
magick lena.png -format "%[fx:mean]" info:
0.502798
```

**Reply by PenghaiZhang** [reply]
Yeah I have tried `fx:mean` it works!

**Comment by PenghaiZhang** [comment]
I just found the thumbnails generated from `mean` are much darker :thinking: .

**Comment by snibgo** [comment]
`image.` was a keyword in the old fx. But the meaning wasn't documented, and it was superfluous, being replaceable by nothing or `u.` or `s.`. So I didn't implement it in the new fx.

> I just found the thumbnails generated from mean are much darker 🤔 .

Darker than what? Please show your command, and sample input and outputs.

**Reply by PenghaiZhang** [reply]
The first image is the original image. The second one is the thumbnail.

The command is `convert xxx.jpg -threshold 99% -format "%[fx:100*mean]" info:`

![Indian-Ring-Necked-Parakeet](https://user-images.githubusercontent.com/47203811/218885267-4c6d6976-2830-4ed0-bfa8-f5f93a9bd79a.jpg)
![629b203e-ba06-4d6f-87e7-263b0f282f66](https://user-images.githubusercontent.com/47203811/218885277-4d1aa99f-99e7-49f3-b572-7cf2702f4811.jpg)

**Reply by PenghaiZhang** [reply]
Do you I need to set Grayscale ?

**Comment by snibgo** [comment]
That command doesn't create a thumbnail, or any image. It merely writes text data to stdout.

**Reply by PenghaiZhang** [reply]
Sorry I copied the wrong one. Should be this one.


`convert -size 600x600 -colorspace RGB a.jpg -thumbnail 500x500> -gravity center b.jpeg`


On my alpine linux where ImageMagick version is 7.1.0.52 the thumbnail is very dark as per above .

On my Ubuntu where the version is much older (7.0.8-25) the colour of thumbnail just looks like the original image.

I am not sure if this is caused by the different version or different OS.

**Comment by snibgo** [comment]
"`-colorspace RGB`" converts the colorspace to linear RGB. I suppose the input was sRGB. Why do you convert it to linear RGB? If you want the output to be sRGB, then remove "`-colorspace RGB`".

In that command, "`-size 600x600`" does nothing and can be removed.

**Reply by PenghaiZhang** [reply]
Thanks @snibgo . The code was written by other developers long long ago and this is their comment for `colorspace RGB`.

> Make sure we don't output CMYK as IE won't show it

**Reply by PenghaiZhang** [reply]
Thanks for your help! After removing `-colorspace RGB` the thumbnails look all good!

**Comment by fmw42** [comment]
Then change -colorspace RGB to -colorspace sRGB.  Long ago, they were swapped, but finally got corrected.

**Comment by snibgo** [comment]
> Make sure we don't output CMYK as IE won't show it

As fmw42 says, you can use "`-colorspace sRGB`" for that, with an `s`.
